Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(0, fpu_get(0) / fpu_get(insn.modrm().register_index()));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f32 = bit_cast<float>(new_f32.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, fpu_get(0) / f32);
+    }
+}
+void SoftFPU::FDIV_RM64(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    } else {
+        auto new_f64 = insn.modrm().read64(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f64 = bit_cast<double>(new_f64.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, fpu_get(0) / f64);
+    }
+}
+void SoftFPU::FDIVP(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+    fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    fpu_pop();
+}
+
+void SoftFPU::FDIVR_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_set(0, fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f32 = bit_cast<float>(new_f32.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, f32 / fpu_get(0));
+    }
+}
+void SoftFPU::FDIVR_RM64(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        // XXX this is FDIVR, Instruction decodes this weirdly
+        //fpu_set(insn.modrm().register_index(), fpu_get(0) / fpu_get(insn.modrm().register_index()));
+        fpu_set(insn.modrm().register_index(), fpu_get(insn.modrm().register_index()) / fpu_get(0));
+    } else {
+        auto new_f64 = insn.modrm().read64(m_cpu, insn);
+        // FIXME: Respect shadow values
+        auto f64 = bit_cast<double>(new_f64.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, f64 / fpu_get(0));
+    }
+}
+void SoftFPU::FDIVRP(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+    fpu_set(insn.modrm().register_index(), fpu_get(0) / fpu_get(insn.modrm().register_index()));
+    fpu_pop();
+}
+
+void SoftFPU::FIDIV_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m16int = (i16)insn.modrm().read16(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    // FIXME: Raise IA on 0 / _=0, raise Z on finite / +-0
```

Comment:
No underflow? Precision (for denormal result)?

Reply:
We don't handle calculation precision anyway
and underflow can mean stack underflow or value underflow, value underflow should not happen by dividing by a non zero int, i think

Reply:
If you divide a denormal by a large integer it should value underflow I think?

Reply:
I think underflow means it would go to -infinity,
undeflow also accurs when trying to store a to small value into an other type, but would need to read again